syntax = "proto3";

service NameNodeService {
  rpc Create(FilePath) returns (Status) {}
  rpc AllocateBlocks(FileData) returns (BlockAllocation) {}
  rpc Append(FileData) returns (Status) {}
  rpc GetBlockLocations(FilePath) returns (BlockLocations) {}
  rpc RegisterDataNode(DataNodeInfo) returns (Status) {}
  rpc DataNodeHeartbeat(DataNodeStatus) returns (Status) {}
  rpc RelocateBlocks(BlockRelocation) returns (Status) {}
}

message FilePath {
  string path = 1;
}

message FileData {
  string name = 1;
  bytes data = 2;
}

message BlockAllocation {
  string name = 1;
  repeated string blockIds = 2;
}

message BlockLocations {
  string name = 1;
  map<string, string> locations = 2;  // Map from block ID to DataNode ID
}

message DataNodeInfo {
  string id = 1;
  string address = 2;
}

message DataNodeStatus {
  string id = 1;
  bool isAlive = 2;
}

message BlockRelocation {
  string blockId = 1;
  string sourceDataNodeId = 2;
  string targetDataNodeId = 3;
}

message Status {
  bool success = 1;
  string message = 2;
}
